src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -3429,62 +3432,113 @@ public String toEngineeringString() {
      * @see #toEngineeringString()
      */
     public String toPlainString() {
-        if(scale==0) {
-            if(intCompact!=INFLATED) {
-                return Long.toString(intCompact);
-            } else {
-                return intVal.toString();
+        int scale = this.scale;
+        if (scale == 0)
+            return unscaledString();
+
+        int signum = signum();
+        if (scale < 0) { // No decimal point
+            if (signum == 0)
+                return "0";
+            int trailingZeros = checkScaleNonZero((-(long) scale));
+            String str = unscaledString();
+            int len = str.length() + trailingZeros;
+            if (len < 0) {
+                throw new OutOfMemoryError("too large to fit in a String");
             }
+            return new StringBuilder(len)
+                    .append(str)
+                    .repeat('0', trailingZeros)
+                    .toString();
         }
-        if(this.scale<0) { // No decimal point
-            if(signum()==0) {
-                return "0";
+
+        long intCompact = this.intCompact;
+        // currency fast path
+        if (intCompact != INFLATED) {
+            long intCompactAbs = Math.abs(intCompact);
+            if (scale == 2 & (int) intCompact == intCompactAbs) { // intCompact >= 0 && intCompact <= Integer.MAX_VALUE
+                return scale2((int) intCompact);
             }
-            int trailingZeros = checkScaleNonZero((-(long)scale));
-            String str = intCompact != INFLATED
-                ? Long.toString(intCompact)
-                : intVal.toString();
-            int len = str.length() + trailingZeros;
+            return getValueString(signum, intCompactAbs, DecimalDigits.stringSize(intCompactAbs), scale);
+        }
+
+        return getValueString(signum, intVal.abs().toString(), scale);
+    }
+
+    private static String getValueString(int signum, long intCompactAbs, int intCompactAbsSize, int scale) {
```

Comment:
I recommend naming this `getCompactValueString` to indicate this differs from `getValueString` by accepting only compact values. Something like:

"Returns a digit.digit string for a compact value"